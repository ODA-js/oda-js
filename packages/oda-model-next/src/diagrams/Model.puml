@startuml model

interface IPackage
interface IPackagedItems

interface IEntity extends IPackagedItems
interface IEnum extends IPackagedItems
interface IEnumItem
interface IField
interface IModel
interface IMutation extends IPackagedItems

together {
  interface IRelation
  interface IBelongsToMany extends IRelation
  interface IBelongsTo extends IRelation
  interface IHasMany extends IRelation
  interface IHasOne extends IRelation
}

IField "1" *--  IRelation : relation
IPackage "1" *-- "*" IPackagedItems : items
IEntity "1" *-- "*" IField: fields
IModel "1" *-- "*" IPackage : packages
IEnum "1" *-- "*" IEnumItem : values

@enduml
