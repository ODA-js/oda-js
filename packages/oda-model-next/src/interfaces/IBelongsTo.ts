import { IEntityRef } from './IEntityRef';
import { IRelation, IRelationInit, IRelationStore } from './IRelation';
import { EntityRefTransform, FieldTransformType } from './types';

export interface IBelongsToInit extends Partial<IRelationInit> {
  belongsTo: string | IEntityRef;
}

export interface IBelongsToStore extends IRelationStore {
  belongsTo: IEntityRef;
}

export interface IRelationTransform {
  belongsTo: EntityRefTransform;
  fields: FieldTransformType;
}

export interface IBelongsTo extends IRelation {
  readonly verb: 'BelongsTo';
  readonly belongsTo: IEntityRef;
}

