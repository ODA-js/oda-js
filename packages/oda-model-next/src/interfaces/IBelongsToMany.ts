import { IEntityRef } from './IEntityRef';
import { IRelation, IRelationInit, IRelationStore } from './IRelation';
import { FieldTransformType, EntityRefTransform } from './types';

export interface IBelongsToManyInit extends Partial<IRelationInit> {
  belongsToMany: string | IEntityRef;
  using: string | IEntityRef;
}

export interface IBelongsToManyStore extends IRelationStore {
  belongsToMany: IEntityRef;
  using: IEntityRef;
}

export interface IRelationTransform {
  belongsToMany: EntityRefTransform;
  using: EntityRefTransform;
  fields: FieldTransformType;
}

export interface IBelongsToMany extends IRelation {
  readonly verb: 'BelongsToMany';
  readonly belongsToMany: IEntityRef;
  readonly using: IEntityRef;
}
