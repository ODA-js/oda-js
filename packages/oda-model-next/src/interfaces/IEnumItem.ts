import { IContextable } from '../contexts/IContextable';
import { IEnumContext } from '../contexts/IEnumContext';
import { IModelType, INamedItem } from './IModelType';

export type EnumInitItem = IEnumItemInit | string;

export interface IEnumItemInit extends INamedItem {
  value?: string;
  type?: string;
}

export interface IEnumItemStore extends INamedItem {
  value: string;
  type: string;
}

export interface IEnumItem extends IModelType, Readonly<IEnumItemInit>, IContextable<IEnumContext> {
  readonly value: string;
  readonly type: string;
}
