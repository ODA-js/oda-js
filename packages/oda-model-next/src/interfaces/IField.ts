import { Map } from 'immutable';

import { IContextable } from '../contexts/IContextable';
import { IEntityContext } from '../contexts/IEntityContext';
import { IRelationContext } from '../contexts/IRelationContext';
import { IEntity } from './IEntity';
import { IEntityRef } from './IEntityRef';
import { FieldArgsInput, IFieldArg, IFieldArgInit } from './IFieldArg';
import { IModelType, INamedItem } from './IModelType';
import { IRelation, IRelationInit } from './IRelation';
import { RelationInit, FieldArgsTransform } from './types';

export interface IFieldACL {
  read: string[];
  update: string[];
}

export interface IFieldMetaData {
  acl: IFieldACL;
}

export interface IFieldStorage {
  derived: boolean;
  persistent: boolean;
  required: boolean;
  indexed: boolean | string | string[];
  identity: boolean | string | string[];
}

export interface IFieldInit
  extends Partial<IFieldStorage>,
    Partial<IFieldMetaData>,
    INamedItem {
  entity?: IEntity;
  type?: string;
  args?: FieldArgsInput;
  order?: number;
  relation?: Partial<RelationInit>;
}

export interface IFieldStore extends IFieldStorage, IFieldMetaData, INamedItem {
  entity: IEntity;
  type: string;
  args?: Map<string, IFieldArg>;
  idKey?: IEntityRef;
  order: number;
  relation?: IRelation;
}

export interface IFieldTransform {
  args: FieldArgsTransform;
  relation: {
    transform: (inp: Partial<IRelationInit>, field: IField) => IRelation;
    reverse: (inp: IRelation) => Partial<IRelationInit>;
  };
}

export interface IField
  extends IModelType,
    IContextable<IEntityContext | IRelationContext> {
  readonly modelType: 'field';
  readonly type: string;
  readonly args?: Map<string, IFieldArgInit>;
  readonly idKey?: IEntityRef;
  readonly order: number;
  readonly relation: IRelation;
  readonly derived: boolean;
  readonly persistent: boolean;
  readonly required: boolean;
  readonly indexed: boolean | string | string[];
  readonly identity: boolean | string | string[];
  readonly acl: Partial<IFieldACL>;
}
