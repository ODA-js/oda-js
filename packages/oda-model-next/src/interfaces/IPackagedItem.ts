import { IContextable } from '../contexts/IContextable';
import { IPackageContext } from '../contexts/IPackageContext';

export type PackagedItemInit = IPackagedItemInit | string;

export interface IPackagedItemInit {
  name: string;
}

export interface IPackagedItemStore {
  name: string;
}

export interface IPackagedItem extends IContextable<IPackageContext> {
  readonly name: string;
  toJS(): object;
}
