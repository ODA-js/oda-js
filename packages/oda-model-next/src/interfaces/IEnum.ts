import { Map } from 'immutable';

import { IContextable } from '../contexts/IContextable';
import { IPackageContext } from '../contexts/IPackageContext';
import { EnumInitItem, IEnumItem, IEnumItemInit } from './IEnumItem';
import { IModelType, INamedItem } from './IModelType';
import { IPackagedItem, IPackagedItemInit, IPackagedItemStore } from './IPackagedItem';

export interface IEnumInit extends INamedItem, IPackagedItemInit {
  values: EnumInitItem[] | {
    [name: string]: EnumInitItem;
  };
}

export interface IEnumStore extends INamedItem, IPackagedItemStore {
  values: Map<string, IEnumItem>;
}

export interface IEnumTransform {
  values: EnumItemTransformType;
}

export interface IEnum extends IModelType, IPackagedItem, IContextable<IPackageContext> {
  readonly modelType: 'enum';
  readonly values: Map<string, IEnumItem>;
}

export type EnumItemTransformType = {
  transform: (input: EnumInitItem[] | {
    [name: string]: EnumInitItem;
  }, en: IEnum) => Map<string, IEnumItem>;
  reverse: (input: Map<string, IEnumItem>) => IEnumItemInit[];
};
