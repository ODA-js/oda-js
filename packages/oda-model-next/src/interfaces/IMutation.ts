import { Map, Set } from 'immutable';

import { IContextable } from '../contexts/IContextable';
import { IPackageContext } from '../contexts/IPackageContext';
import { ArrayToSet } from '../model/utils';
import { FieldArgsInput, IFieldArg, IFieldArgInit } from './IFieldArg';
import { IModelType, INamedItem } from './IModelType';
import { IPackagedItem, IPackagedItemInit, IPackagedItemStore } from './IPackagedItem';
import { FieldArgsTransform } from './types';

export interface IMutationACL {
  execute: string[];
}

export interface IMutationMetaData {
  acl?: IMutationACL;
}

export interface IMutationACLStore {
  execute: Set<string>;
}

export interface IMutationMetaDataStore {
  acl?: IMutationACLStore;
}

export interface IMutationStore extends IMutationMetaDataStore, INamedItem, IPackagedItemStore {
  args: Map<string, IFieldArg>;
  payload: Map<string, IFieldArg>;
}

export interface IMutationInit extends Partial<IMutationMetaData>, INamedItem, IPackagedItemInit {
  args: FieldArgsInput;
  payload: FieldArgsInput;
}

export interface IMutationACLTransform {
  execute: ArrayToSet<string>;
}

export interface IMutationTransform {
  args: FieldArgsTransform;
  payload: FieldArgsTransform;
  acl: IMutationACLTransform;
}

export interface IMutation extends IModelType, IPackagedItem, IContextable<IPackageContext> {
  readonly modelType: 'mutation';
  readonly args: Map<string, IFieldArgInit>;
  readonly payload: Map<string, IFieldArgInit>;
  readonly acl?: IMutationACLStore;
}

