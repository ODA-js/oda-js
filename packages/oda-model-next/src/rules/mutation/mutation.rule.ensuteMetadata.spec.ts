import 'jest';
import { IPackagedItemInit } from '../../interfaces/IPackagedItem';
import { Entity } from '../../model/Entity';
import { Model } from '../../model/Model';
import { Package } from '../../model/Package';
import Rule from './ensureMetadata';
import { IMutationInit } from '../../interfaces/IMutation';

describe('rule', () => {
  let rule: Rule;

  beforeEach(() => {
    rule = new Rule();
  });

  it('fix acl', () => {
    let update = {};
    const updateWith = jest.fn(args => {
      update = {
        ...update,
        ...args,
      };
    });

    const result = rule.validate({
      model: {},
      package: {},
      mutation: {
        name: 'mutation',
        args: {
          input: {},
        },
        payload: {
          result: {},
        },
        updateWith,
      },
    } as any);
    expect(update).toMatchSnapshot();
  });
});
