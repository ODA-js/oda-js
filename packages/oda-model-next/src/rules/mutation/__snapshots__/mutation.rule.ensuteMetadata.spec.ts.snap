// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule fix acl 1`] = `
Object {
  "acl": Object {
    "execute": Array [],
  },
}
`;
